// vue-snack/dist/vue-snack.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
	typeof define === 'function' && define.amd ? define(factory) :
	(global.VueSnackPlugin = factory());
}(this, (function () { 'use strict';

function ___$insertStyle(css) {
  if (!css) {
    return;
  }
  if (typeof window === 'undefined') {
    return;
  }

  var style = document.createElement('style');

  style.setAttribute('type', 'text/css');
  style.innerHTML = css;
  document.head.appendChild(style);

  return css;
}

___$insertStyle("/* latin */\n@font-face {\n  font-family: 'Roboto';\n  font-style: normal;\n  font-weight: 400;\n  src: local(\"Roboto\"), local(\"Roboto-Regular\"), url(\"./KFOmCnqEu92Fr1Mu4mxK.woff2\") format(\"woff2\");\n  unicode-range: U+0000-00FF, U+0131, U+0152-0153, U+02BB-02BC, U+02C6, U+02DA, U+02DC, U+2000-206F, U+2074, U+20AC, U+2122, U+2191, U+2193, U+2212, U+2215, U+FEFF, U+FFFD; }\n\n/* latin */\n@font-face {\n  font-family: 'Roboto';\n  font-style: normal;\n  font-weight: 500;\n  src: local(\"Roboto Medium\"), local(\"Roboto-Medium\"), url(\"./KFOlCnqEu92Fr1MmEU9fBBc4.woff2\") format(\"woff2\");\n  unicode-range: U+0000-00FF, U+0131, U+0152-0153, U+02BB-02BC, U+02C6, U+02DA, U+02DC, U+2000-206F, U+2074, U+20AC, U+2122, U+2191, U+2193, U+2212, U+2215, U+FEFF, U+FFFD; }\n\n.snackbar {\n  --primary: #C849F4;\n  --background: #353535;\n  --text: #E3E3E3;\n  --font: 'Roboto', sans-serif;\n  --times: 1rem;\n  --padd: 1rem; }\n\n.snackbar {\n  font-size: 1rem;\n  background-color: var(--background);\n  color: var(--text);\n  font-family: var(--font);\n  display: -webkit-box;\n  display: -ms-flexbox;\n  display: flex;\n  -webkit-box-align: center;\n      -ms-flex-align: center;\n          align-items: center;\n  -webkit-box-pack: center;\n      -ms-flex-pack: center;\n          justify-content: center;\n  position: fixed;\n  min-height: var(--times);\n  padding: var(--padd);\n  border-radius: 4px; }\n  @media screen and (max-width: 576px) {\n    .snackbar {\n      border-radius: 0;\n      width: calc(100% - (2 * var(--padd))); } }\n  @media screen and (min-width: 576px) {\n    .snackbar.top, .snackbar.bottom {\n      -webkit-transform: translateX(-50%);\n              transform: translateX(-50%);\n      left: 50%; }\n      .snackbar.top-left, .snackbar.bottom-left {\n        left: 1rem; }\n      .snackbar.top-right, .snackbar.bottom-right {\n        right: 1rem; } }\n  .snackbar[class*=\"bottom\"] {\n    bottom: 1rem; }\n  .snackbar[class*=\"top\"] {\n    top: 1rem; }\n  @media screen and (max-width: 576px) {\n    .snackbar {\n      left: 0; }\n      .snackbar[class*=\"bottom\"] {\n        bottom: 0; }\n      .snackbar[class*=\"top\"] {\n        top: 0; } }\n  .snackbar .snackbar__text {\n    -webkit-box-flex: 1;\n        -ms-flex: 1;\n            flex: 1; }\n  .snackbar .snackbar__action {\n    text-transform: uppercase;\n    font-weight: 500;\n    margin: 0 0 0 1rem;\n    color: var(--primary);\n    cursor: pointer;\n    -webkit-user-select: none;\n       -moz-user-select: none;\n        -ms-user-select: none;\n            user-select: none; }\n  .snackbar .times {\n    cursor: pointer;\n    display: -webkit-box;\n    display: -ms-flexbox;\n    display: flex;\n    -webkit-box-align: center;\n        -ms-flex-align: center;\n            align-items: center;\n    -webkit-box-pack: center;\n        -ms-flex-pack: center;\n            justify-content: center;\n    height: var(--times);\n    width: var(--times);\n    margin-left: 1rem;\n    -webkit-transform: scale(1.25);\n            transform: scale(1.25);\n    position: relative; }\n    .snackbar .times:before {\n      display: block;\n      content: '';\n      height: 100%;\n      width: 100%;\n      position: absolute;\n      top: 0;\n      left: 0;\n      -webkit-transform: scale(0.9);\n              transform: scale(0.9);\n      border-radius: 50%;\n      -webkit-transition: -webkit-transform .1s ease-in;\n      transition: -webkit-transform .1s ease-in;\n      transition: transform .1s ease-in;\n      transition: transform .1s ease-in, -webkit-transform .1s ease-in; }\n    .snackbar .times:hover:before {\n      background-color: rgba(255, 255, 255, 0.2);\n      -webkit-transform: scale(2);\n              transform: scale(2); }\n\n.snack-enter-active, .snack-leave-active {\n  -webkit-transition: all .4s ease;\n  transition: all .4s ease; }\n\n.snack-enter, .snack-leave-to {\n  opacity: 0; }\n  @media screen and (min-width: 576px) {\n    .snack-enter.bottom, .snack-leave-to.bottom {\n      -webkit-transform: translateY(100%) translateX(-50%);\n              transform: translateY(100%) translateX(-50%); }\n    .snack-enter.top, .snack-leave-to.top {\n      -webkit-transform: translateY(-100%) translateX(-50%);\n              transform: translateY(-100%) translateX(-50%); } }\n  @media screen and (max-width: 576px) {\n    .snack-enter[class*=\"top\"], .snack-leave-to[class*=\"top\"] {\n      -webkit-transform: translateY(-100%);\n              transform: translateY(-100%); } }\n  @media screen and (max-width: 576px) {\n    .snack-enter[class*=\"bottom\"], .snack-leave-to[class*=\"bottom\"] {\n      -webkit-transform: translateY(100%);\n              transform: translateY(100%); } }\n  @media screen and (min-width: 576px) {\n    .snack-enter[class*=\"-left\"], .snack-leave-to[class*=\"-left\"] {\n      -webkit-transform: translateX(-50%);\n              transform: translateX(-50%); } }\n  @media screen and (min-width: 576px) {\n    .snack-enter[class*=\"-right\"], .snack-leave-to[class*=\"-right\"] {\n      -webkit-transform: translateX(50%);\n              transform: translateX(50%); } }\n");

var Snack = { render: function render() {
    var _vm = this;var _h = _vm.$createElement;var _c = _vm._self._c || _h;return _c('transition', { attrs: { "name": "snack" } }, [_vm.active ? _c('div', { staticClass: "snackbar", class: _vm.config.position, style: _vm.styles }, [_c('div', { staticClass: "snackbar__text" }, [_vm._v(_vm._s(_vm.text))]), _vm._v(" "), _vm.button ? [_c('div', { staticClass: "snackbar__action", on: { "click": function click($event) {
          $event.preventDefault();return _vm.action($event);
        } } }, [_vm._v(_vm._s(_vm.button))])] : _vm._e(), _vm._v(" "), _vm.config.close ? [_c('div', { staticClass: "times", on: { "click": function click($event) {
          _vm.$emit('close');
        } } }, [_c('svg', { attrs: { "xmlns": "http://www.w3.org/2000/svg", "width": "24", "height": "24", "viewBox": "0 0 24 24" } }, [_c('path', { attrs: { "fill": "#E3E3E3", "d": "M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z" } }), _vm._v(" "), _c('path', { attrs: { "d": "M0 0h24v24H0z", "fill": "none" } })])])] : _vm._e()], 2) : _vm._e()]);
  }, staticRenderFns: [],
  name: 'Snackbar',
  props: {
    config: Object
  },
  data: function data(_) {
    return {
      active: false,
      text: '',
      button: '',
      theme: 'default',
      action: null
    };
  },
  computed: {
    styles: function styles() {
      return "--primary: " + this.config[this.theme].primary;
    }
  }
};

var Default = {
  default: {
    primary: '#2C89F1'
  },
  success: {
    primary: '#00DEB2'
  },
  danger: {
    primary: '#FF0057'
  },
  background: '#353535',
  text: '#E3E3E3',
  time: 7500,
  position: 'bottom',
  close: false
};

//
// vue PlugIn
//
var Constructor = void 0;
var component = null;
var timeout = null;
var closing = null;
var options = null;
var config = {};
var methods = {
  danger: 'danger',
  show: 'default',
  success: 'success'
};

var sleep = function sleep(time) {
  return new Promise(function (resolve, reject) {
    return setTimeout(function (_) {
      return resolve();
    }, time);
  });
};

var create = function create(parent) {
  document.addEventListener('DOMContentLoaded', function (event) {
    var div = document.createElement('div');
    div.id = 'snackbar-' + Date.now();
    document.body.appendChild(div);
    config.config = Default;
    component = new Constructor({
      propsData: config
      // parent
    });
    component.$on('close', function (_) {
      return close();
    });
    component.$mount('#' + div.id);
  });
};

var close = function close() {
  component.active = false;
  clearTimeout(timeout);
};

var actions = async function actions(params, theme) {
  options = params;
  if (component.active && closing) {
    return;
  }
  if (component.active) {
    close();
    closing = true;
    await sleep(400);
    closing = false;
  }
  if (typeof options === 'string') {
    options = { text: options };
  }
  var fn = options.action;
  options.action = async function () {
    if (!fn) {
      return close();
    }
    fn();
    close();
  };

  Object.assign(config.config, Default[theme]);
  Object.assign(component, Object.assign({}, options, { theme: theme }));
  component.active = true;
  timeout = setTimeout(close, Default.time);
};

var $snack = function $snack(opt) {
  var news = {};
  var themes = {};
  opt.methods = opt.methods || [];
  opt.methods.forEach(function (item) {
    news[item.name] = item.name;
    themes[item.name] = {
      primary: item.color || Default.default.primary
    };
  });
  Object.assign(Default, themes);
  var all = {};
  var meth = Object.assign({}, methods, news);

  var _loop = function _loop(m) {
    all[m] = function (params) {
      return actions(params, meth[m]);
    };
  };

  for (var m in meth) {
    _loop(m);
  }
  return all;
};

var plugin$1 = {
  install: function install(Vue) {
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    Constructor = Vue.extend(Snack);
    create();
    Object.assign(Default, options);
    Vue.prototype.$snack = $snack(options);
  }
};

// Auto-install
var _Vue = null;
if (typeof window !== 'undefined') {
  _Vue = window.Vue;
} else if (typeof global !== 'undefined') {
  _Vue = global.Vue;
}
if (_Vue) {
  _Vue.use(plugin$1);
}

return plugin$1;

})));
